nav2_costmap_2d/src/costmap_2d.cpp

```diff
@@ -399,29 +399,59 @@ void Costmap2D::updateOrigin(double new_origin_x, double new_origin_y)
 bool Costmap2D::setConvexPolygonCost(
   const std::vector<geometry_msgs::msg::Point> & polygon,
   unsigned char cost_value)
+{
+  std::vector<std::pair<MapLocation, unsigned char>> polygon_map_region;
+  if (getMapRegionOccupiedByPolygon(polygon, polygon_map_region)) {
+    return false;
+  }
+
+  // set the cost of those cells
+  setMapRegionOccupiedByPolygon(polygon_map_region, cost_value);
+  return true;
+}
+
+void Costmap2D::setMapRegionOccupiedByPolygon(
+  const std::vector<std::pair<MapLocation, unsigned char>> & polygon_map_region,
+  unsigned char new_cost_value)
+{
+  for (const auto & [cell, _] : polygon_map_region) {
+    setCost(cell.x, cell.y, new_cost_value);
+  }
+}
+
+void Costmap2D::restoreMapRegionOccupiedByPolygon(
+  const std::vector<std::pair<MapLocation, unsigned char>> & polygon_map_region)
+{
+  for (const auto & [cell, cost_value] : polygon_map_region) {
+    setCost(cell.x, cell.y, cost_value);
+  }
+}
+
+bool Costmap2D::getMapRegionOccupiedByPolygon(
+  const std::vector<geometry_msgs::msg::Point> & polygon,
+  std::vector<std::pair<MapLocation, unsigned char>> & polygon_map_region)
 {
   // we assume the polygon is given in the global_frame...
   // we need to transform it to map coordinates
   std::vector<MapLocation> map_polygon;
-  for (unsigned int i = 0; i < polygon.size(); ++i) {
+  for (const auto & cell : polygon) {
     MapLocation loc;
-    if (!worldToMap(polygon[i].x, polygon[i].y, loc.x, loc.y)) {
+    if (!worldToMap(cell.x, cell.y, loc.x, loc.y)) {
       // ("Polygon lies outside map bounds, so we can't fill it");
       return false;
     }
     map_polygon.push_back(loc);
   }
 
-  std::vector<MapLocation> polygon_cells;
-
   // get the cells that fill the polygon
+  std::vector<MapLocation> polygon_cells;
```

Comment:
Do a `polygon_cells.reserve(100);` to reserve the size to 100 cells, that'll help with resizings as you're pushing back. Larger footprints will still need to resize, but smaller ones shouldn't.